src/tyr/route_serializer_osrm.cc

```diff
@@ -283,23 +206,85 @@ json::MapPtr geojson_shape(const std::vector<PointLL> shape) {
   return geojson;
 }
 
+// Maps Valhalla edge and road class to an OpenLR form-of-way.
+FormOfWay RoadClassToFOW(const TripLeg::Node& node) {
+  const bool is_missing_rc = !node.has_edge() || !node.edge().has_road_class();
+  if (is_missing_rc || !node.edge().has_use()) {
+    return FormOfWay::OTHER;
+  }
+  const TripLeg::Use use = node.edge().use();
+  const valhalla::RoadClass rc = node.edge().road_class();
+  const bool one_way =
+      node.edge().has_traversability() && node.edge().traversability() != TripLeg::kBoth;
+  if (node.edge().roundabout()) {
+    return FormOfWay::ROUNDABOUT;
+  } else if (use == TripLeg::kRampUse || use == TripLeg::kTurnChannelUse) {
+    return FormOfWay::SLIPROAD;
+  } else if (rc == valhalla::RoadClass::kMotorway && one_way) {
+    return FormOfWay::MOTORWAY;
+  } else if (rc <= valhalla::RoadClass::kTertiary && one_way) {
+    return FormOfWay::MULTIPLE_CARRIAGEWAY;
+  } else if (rc <= valhalla::RoadClass::kTertiary) {
+    return FormOfWay::SINGLE_CARRIAGEWAY;
+  } else {
+    return FormOfWay::OTHER;
+  }
+}
+
+// TODO(mookerji): Move this shared header file somewhere and remove also from tests/openlr.cc
+std::string encode64(const std::string& val) {
+  using namespace boost::archive::iterators;
+  using It = base64_from_binary<transform_width<std::string::const_iterator, 6, 8>>;
+  auto tmp = std::string(It(std::begin(val)), It(std::end(val)));
+  return tmp.append((3 - val.size() % 3) % 3, '=');
+}
+
+// TODO(mookerji): comment
+std::string openlr_shape(const valhalla::TripRoute& route) {
```

Comment:
the first and last edges of a leg may (and probably will) have a point that is not at a graph node (intersection) at the beginning or end of the edge respectively. i think the canonical way to handle such openlrs is to actually use the point at the preceding or following intersection respectively and then set positive offset (poff) or negative offset (noff) respectively. i dont think we need to handle that in the first go round here. but in the future it might make sense, in triplegbuilder, to add some information to the leg abotu the interection the edge starts/ends on as opposed to the point the route starts/ends on.